sensors/gps/pa6h.c

```diff
@@ -41,8 +52,274 @@ typedef struct {
+
+	if (ret >= 0) {
+		/* update position and nul terminate the buffer */
+		pos += ret;
+		buf[pos] = '\0';
+
+		/* find beginning of first message */
+		tokenStart = strchr(buf, '$');
+		tokenEnd = tokenStart;
+		while (tokenStart != NULL) {
+			remStart = tokenStart;
+
+			/* Seek current message`s end. Avoid multiple search for the same 'tokenEnd' if 'tokenStart' is behind it */
+			if (tokenStart >= tokenEnd) {
+				tokenEnd = strchr(tokenStart, '*');
+				if (tokenEnd == NULL) {
+					break;
+				}
+				nextStartToken = strchr(tokenEnd, '$');
+			}
+
+			/* Ensure space for checksum and nul character not interfering with next message. Lack '/r/n' at the end is discarded */
+			len = ((nextStartToken == NULL) ? &buf[pos] : nextStartToken) - tokenEnd;
+			if (len < 3) {
+				break;
+			}
+
+			val = strtol(tokenEnd + 1, (char **)NULL, 16);
+			if (!(val == 0 && errno == EINVAL) && val < 0xff) {
+				len = tokenEnd - tokenStart;
+
+				/* calculate checksum */
+				for (i = 1, checksum = 0; i <= len - 1; i++) {
+					checksum ^= tokenStart[i];
+				}
+
+				/* Save message if checksum matches and there is enough space */
+				if (checksum == val && inboxFill < inboxSz) {
+					inbox[inboxFill].msg = tokenStart;
+					inbox[inboxFill].sz = len + 4; /* +1 for '*', +2 for checksum, +1 for nul */
+					inboxFill++;
+				}
+			}
+
+			/* find next starting token */
+			tokenStart = strchr(tokenStart + 1, '$');
+		}
+	}
+
+	/* Discard buffer remainings if last message was complete */
+	if (tokenStart == NULL && tokenEnd != NULL) {
+		remStart = buf;
+		pos = 0;
+	}
+
+	for (i = 0; i < inboxFill; i++) {
+		inbox[i].msg[inbox[i].sz] = 0;
+	}
+
+	return inboxFill;
+}
+
+
+/*
+* With assumption that at buf[0] is '$' validates if message is pmtk acknowledgment message coming from msgId message.
+* Returns status flag if message is pmtk ack and source id matches `msgId`. Returns -1 otherwise.
+*/
+static int pa6h_pmtkAckCheck(const char *buf, size_t bufSz, unsigned int msgId)
+{
+	/*
+	* Pmtk ack message template:
+	* $PMTK001,cmd,o*cs
+	*
+	* cmd - three digit source command id
+	* o   - one digit output
+	* cs  - two digit checksum
+	*
+	* message length is 17 bytes, all buffers are +1 for nul character
+	*/
+	unsigned int val, i;
+	char checksum;
+
+	if (bufSz <= PMTK_ACK_MSG_LEN || msgId > 999 || msgId == 0) {
+		return -1;
+	}
+
+	/* validate general static tokens. Assures that atoi/strtod will work on data chunks of correct length/semantics */
+	if (buf[0] != '$' || buf[8] != ',' || buf[12] != ',' || buf[14] != '*') {
+		return -1;
+	}
+
+	/* checksum validation */
+	val = strtol(&buf[15], (char **)NULL, 16);
+	if ((val == 0 && errno == EINVAL) || val > 0xff) {
+		return -1;
+	}
+
+	for (i = 1, checksum = 0; i < 14; i++) {
+		checksum ^= buf[i];
+	}
+
+	if (checksum != val) {
+		printf("%x %x\n", checksum, val);
+		return -1;
+	}
+
+	/* ACK message header check */
+	if (strncmp(&buf[0], PMTK_HEADER, sizeof(PMTK_HEADER) - 1) != 0) {
+		return -1;
+	}
```

Comment:
Implemented, but with additional "$" check. It is somehow redundant to prior `pa6h_receive` check, but more atomic: this function checks the full pmtk message integrity now. No matter if it would be used in `pa6h_receive` or not